File: src/gateway/sse.py
# SSE 流式输出：将回复按标点分句输出，兼容 OpenAI chat.completion.chunk 格式。

import json
import re
import time
from typing import AsyncGenerator
from uuid import uuid4


def split_sentences(text: str) -> list[str]:
    """按中文标点分句，保留标点符号。"""
    parts = re.split(r"([，。？！；,\.!\?;])", text)
    sentences: list[str] = []
    current = ""
    for part in parts:
        current += part
        if re.match(r"[，。？！；,\.!\?;]", part):
            if current.strip():
                sentences.append(current)
            current = ""
    if current.strip():
        sentences.append(current)
    return sentences or [text]


async def generate_sse(
    reply: str,
    *,
    model: str,
    created: int | None = None,
    completion_id: str | None = None,
) -> AsyncGenerator[str, None]:
    """生成 SSE 流：按句逐条输出 delta chunk，最后发送 [DONE]。"""
    created_at = created or int(time.time())
    response_id = completion_id or f"chatcmpl-{uuid4().hex[:12]}"
    for sentence in split_sentences(reply):
        chunk = {
            "choices": [{"delta": {"content": sentence}, "finish_reason": None, "index": 0}],
            "object": "chat.completion.chunk",
            "model": model,
            "created": created_at,
            "id": response_id,
        }
        yield f"data: {json.dumps(chunk, ensure_ascii=False)}\n\n"

    final = {
        "choices": [{"delta": {"content": ""}, "finish_reason": "stop", "index": 0}],
        "object": "chat.completion.chunk",
        "model": model,
        "created": created_at,
        "id": response_id,
    }
    yield f"data: {json.dumps(final, ensure_ascii=False)}\n\n"
    yield "data: [DONE]\n\n"

File: internal/gateway/sse.go
package gateway

import (
	"encoding/json"
	"fmt"
	"io"
	"regexp"
	"strings"
	"time"

	"github.com/google/uuid"
)

var sentenceDelimiter = regexp.MustCompile(`([，。？！；,\.!\?;])`)

type SSEChunk struct {
	Choices []SSEChoice `json:"choices"`
	Object  string      `json:"object"`
	Model   string      `json:"model"`
	Created int64       `json:"created"`
	ID      string      `json:"id"`
}

type SSEChoice struct {
	Delta        SSEDelta `json:"delta"`
	FinishReason *string  `json:"finish_reason"`
	Index        int      `json:"index"`
}

type SSEDelta struct {
	Content string `json:"content"`
}

func SplitSentences(text string) []string {
	parts := sentenceDelimiter.Split(text, -1)
	matches := sentenceDelimiter.FindAllString(text, -1)
	sentences := make([]string, 0, len(parts))
	current := ""
	for i, part := range parts {
		current += part
		if i < len(matches) {
			current += matches[i]
			if current != "" {
				sentences = append(sentences, current)
			}
			current = ""
		}
	}
	if current != "" {
		sentences = append(sentences, current)
	}
	if len(sentences) == 0 {
		return []string{text}
	}
	return sentences
}

func NewCompletionID() string {
	return fmt.Sprintf("chatcmpl-%s", strings.ReplaceAll(uuid.NewString(), "-", "")[:12])
}

func WriteChatCompletionSSE(w io.Writer, reply string, model string, created int64, completionID string) error {
	if created == 0 {
		created = time.Now().Unix()
	}
	if completionID == "" {
		completionID = NewCompletionID()
	}
	for _, sentence := range SplitSentences(reply) {
		chunk := SSEChunk{
			Choices: []SSEChoice{{Delta: SSEDelta{Content: sentence}, FinishReason: nil, Index: 0}},
			Object:  "chat.completion.chunk",
			Model:   model,
			Created: created,
			ID:      completionID,
		}
		if err := writeSSEData(w, chunk); err != nil {
			return err
		}
	}
	stop := "stop"
	final := SSEChunk{
		Choices: []SSEChoice{{Delta: SSEDelta{Content: ""}, FinishReason: &stop, Index: 0}},
		Object:  "chat.completion.chunk",
		Model:   model,
		Created: created,
		ID:      completionID,
	}
	if err := writeSSEData(w, final); err != nil {
		return err
	}
	_, err := io.WriteString(w, "data: [DONE]\n\n")
	return err
}

func writeSSEData(w io.Writer, payload any) error {
	encoded, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	_, err = fmt.Fprintf(w, "data: %s\n\n", encoded)
	return err
}
